packages/regenesis-surgery/scripts/surgery.ts

```diff
@@ -0,0 +1,346 @@
+          [token0, token1, fee]
+        )
+      ),
+      initCodeHash
+    )
+    .toLowerCase()
+}
+
+const toHex32 = (val: string | number | ethers.BigNumber) => {
+  return (
+    '0x' + ethers.BigNumber.from(val).toHexString().slice(2).padStart(64, '0')
+  )
+}
+
+const getMappingKey = (keys: any[], slot: number) => {
+  // TODO: assert keys.length > 0
+  let key = ethers.utils.keccak256(
+    ethers.utils.hexConcat([toHex32(keys[0]), toHex32(slot)])
+  )
+  if (keys.length > 1) {
+    for (let i = 1; i < keys.length; i++) {
+      key = ethers.utils.keccak256(
+        ethers.utils.hexConcat([toHex32(keys[i]), key])
+      )
+    }
+  }
+  return key
+}
+
+const main = async () => {
+  // Load required enviorment variables
+  dotenv.config()
+  const REGEN__STATE_DUMP_FILE = process.env.REGEN__STATE_DUMP_FILE
+  const REGEN__L2_PROVIDER_URL = process.env.REGEN__L2_PROVIDER_URL
+  const REGEN__ETHERSCAN_KEY = process.env.REGEN__ETHERSCAN_KEY
+
+  // Load the state dump from the JSON file
+  const dump: StateDump = JSON.parse(
+    fs.readFileSync(
+      path.resolve(__dirname, `../dumps/${REGEN__STATE_DUMP_FILE}`),
+      { encoding: 'utf8' }
+    )
+  )
+
+  // Set up the L2 provider.
+  const l2Provider = new ethers.providers.JsonRpcProvider(
+    REGEN__L2_PROVIDER_URL
+  )
+
+  // Create an empty object that represents the new genesis state
+  // We're going to move items from the dump into this genesis state
+  const genesis: ChainState = {}
+
+  // Step 1. Transfer the state of each precompiled contract.
+  for (const [address, account] of Object.entries(dump.accounts)) {
+    if (address.startsWith('0x00000000000000000000000000000000000000')) {
+      genesis[address] = account
+      delete dump.accounts[address]
+    }
+  }
+
+  // Step 2. Transfer over each EOA address and turn it into a normal EOA.
+  for (const [address, account] of Object.entries(dump.accounts)) {
+    if (EOA_CODE_HASHES.includes(account.codeHash)) {
+      genesis[address] = {
+        balance: account.balance,
+        nonce: account.nonce,
+        root: DEFAULT_STORAGE_ROOT,
+        codeHash: DEFAULT_CODE_HASH,
+      }
+      delete dump.accounts[address]
+    }
+  }
+
+  /* --- BEGIN UNISWAP SURGERY SECTION --- */
+
+  // Set up the uniswap factory contract reference
+  const UniswapV3Factory = new ethers.Contract(
+    UNISWAP_V3_FACTORY_ADDRESS,
+    UNISWAP_V3_FACTORY_ABI,
+    l2Provider
+  )
+
+  // Step 3. (UNISWAP) Fix the UniswapV3Factory `owner` address.
+  const oldOwnerSlot = toHex32(0)
+  const newOwnerSlot = toHex32(3)
+  dump.accounts[UNISWAP_V3_FACTORY_ADDRESS].storage[newOwnerSlot] =
+    dump.accounts[UNISWAP_V3_FACTORY_ADDRESS].storage[oldOwnerSlot]
+  delete dump.accounts[UNISWAP_V3_FACTORY_ADDRESS].storage[oldOwnerSlot]
+
+  // Step 4. (UNISWAP) Fix the UniswapV3Factory `feeAmountTickSpacing` mapping.
+  const feeEvents = await UniswapV3Factory.queryFilter(
+    UniswapV3Factory.filters.FeeAmountEnabled()
+  )
```

Comment:
improvement here and on top (https://github.com/ethereum-optimism/optimism/pull/1481/files#diff-de41f93baec1842678463433ac56cf5ca6f669d64046729dfbf03dc6b3f03dfeR111-R113) is to use a quorum of providers to ensure this is the correct events